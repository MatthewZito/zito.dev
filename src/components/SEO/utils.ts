import removeMd from 'remove-markdown';

import type { ISeoData, IAbridgedPost } from './types';
import type { IPost, ISiteMetadata } from '@/types';

/**
 * @summary Generate abridged post data from a given `allMdx` edge
 */
export const generatePostData = (post: IPost): IAbridgedPost => {
	const {
		coverImageUrl,
		coverImageAlt,
		datePublished,
		dateModified,
		description,
		title,
		category,
		tags,
		internalContent,
		excerpt
	} = post;

	if (!internalContent) {
		throw Error(
			'[generatePostData] Post does not contain internal content necessary for Rich Tags.'
		);
	}

	const body = removeMd(internalContent);

	return {
		title,
		description: description || excerpt,
		coverImageUrl,
		coverImageAlt,
		datePublished,
		dateModified,
		category: category || 'None',
		tags: tags || [],
		body,
		url: post.url
	};
};

/**
 * @summary Generate SEO metadata
 */
export const generateSeoData = (
	siteMetadata: ISiteMetadata,
	postData?: IAbridgedPost
): ISeoData => {
	const isPost = !!postData;
	const title = postData ? postData.title : siteMetadata.title;
	const type = postData ? 'post' : 'page';
	const imageUrl = postData ? postData.coverImageUrl : siteMetadata.logoUrl;
	const imageAlt = postData ? postData.coverImageAlt : siteMetadata.description;
	const url = postData ? postData.url : siteMetadata.url;
	const description = postData ?
		postData.description :
		siteMetadata.description;

	return {
		isPost,
		type,
		title,
		imageUrl,
		imageAlt,
		url,
		description
	};
};
