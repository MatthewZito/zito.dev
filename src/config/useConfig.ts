import { useStaticQuery, graphql } from 'gatsby';

import type { ISiteConfig } from '@/types';

interface IUserConfigQuery {
	site?: {
		siteMetadata?: {
			config?: ISiteConfig;
		};
	};
}

export const useConfig = (): ISiteConfig => {
	const data = useStaticQuery<IUserConfigQuery>(
		graphql`
			query UserConfig {
				site {
					siteMetadata {
						config {
							contentDir
							assetDir
							embeddedImageWidth
							embeddedVideoWidth
							basePath
							iconPath
							iconCachePaths
							iconList {
								src
								sizes
								type
								purpose
							}
							organization {
								description
								logoUrl
								name
								url
							}
							pathPrefix
							user {
								about
								avatar
								firstName
								github
								email
								id
								surname
								linkedIn
								location
								twitterHandle
							}
							site {
								backgroundColor
								copyright
								description
								language
								facebookAppId
								googleAnalyticsId
								disqusShortname
								logoUrl
								name
								rss
								rssTitle
								themeColor
								title
								titleAbridged
								twitterHandle
								url
							}
						}
					}
				}
			}
		`
	);

	const config = data?.site?.siteMetadata?.config;

	if (!config) throw Error('[useConfig] Failed to query Site Config.');

	return config;
};
