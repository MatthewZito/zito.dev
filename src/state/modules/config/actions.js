/**
 * @description Add Vue Router route object to cache list, to be used by keep-alive
 * @param {object} view The Vue Router object to be cached
 */
export const addViewToCache = ({ commit }, view) => {
  commit('addViewToCache', view);
};

/**
 * @description Remove Vue Router route object from cache list
 * @param {object} view The Vue Router object to be cache-invalidated
 */
export const removeViewFromCache = ({ commit }, view) => {
  commit('removeViewFromCache', view);
};

/**
 * @description Invalidate entire view cache
 */
export const resetViewCache = ({ commit }) => {
  commit('resetViewCache');
};

/**
 * @description Reset Config state to defaults
 */
export const resetState = ({ commit }) => {
  commit('resetState');
};
