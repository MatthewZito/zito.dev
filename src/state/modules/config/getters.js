/**
 * @type {boolean}
 * @summary Get the current nav drawer state
 */
export const isNavDrawerClosed = (state) => !state.isNavDrawerOpen;

/**
 * @type {array}
 * @summary Get all cached views
 */
export const getCachedViews = (state) => state.cachedViews;

/**
 * @type {string}
 * @summary Get app locale / system name
 */
export const appName = (state) => state.appName;

/**
 * @type {string}
 * @summary Get app author
 */
export const appAuthor = (state) => state.author;
