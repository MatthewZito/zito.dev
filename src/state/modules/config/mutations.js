import initState, { cacheState } from './state';

export const addViewToCache = (state, view) => {
  if (state.cachedViews.includes(view?.name)) return;
  if (view?.meta?.cache) {
    state.cachedViews.push(view.name);
  }
};

export const removeViewFromCache = (state, view) => {
  const idx = state.cachedViews.indexOf(view.name);
  idx > -1 && state.cachedViews.splice(idx, 1);
};

export const resetViewCache = state => {
  Object.assign(state, cacheState());
};

export const resetState = state => {
  Object.assign(state, initState());
};
