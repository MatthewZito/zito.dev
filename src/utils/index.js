import store from '@/state';

const { dateFmt } = store.state.config;

/**
 * @summary Standard no-op function
 */
const noop = () => {};

/**
 * @summary Pipe input through N functions
 * @param {function} fn
 * @param  {...function} fns
 * @returns {function} base reducer
 */
const pipe = (fn, ...fns) =>
	(...args) => fns.reduce((acc, fn) => fn(acc), fn(...args));

/**
 * @summary Convert data to blob
 * @param {any} data
 * @returns {Blob}
 */
const blobify = data => new Blob([JSON.stringify(data)],
	{ type : 'application/json; charset=UTF-8' });

/**
 * @summary Convert date to readable string with app-specified format
 * @param {string|Date} raw
 * @returns {string}
 */
const dateConv = raw => Intl.DateTimeFormat('en-US', dateFmt)
	.format(new Date(raw));

/**
 * @summary Convert a date into a UNIX timestamp
 * @param {string|Date} raw
 * @returns {number}
 */
const toUnixTs = raw => new Date(raw)?.getTime() || 0;

/**
 * @summary Is the app running in a local development mode?
 */
const isDev = !!import.meta.env.VITE_LOCAL || import.meta.env.DEV;

/**
 * @summary The current app base URL
 * where local mode is set with a string literal and not an env var
 * because of Vercel CLI modesparadox theatrer
 */
const baseUrl = import.meta.env.VITE_LOCAL
	? 'http://localhost:3000'
	: import.meta.env.VITE_APP_URL_UI;

export {
	noop,
	pipe,
	blobify,
	dateConv,
	toUnixTs,
	isDev,
	baseUrl
};
