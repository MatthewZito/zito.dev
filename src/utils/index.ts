import { IGetMdxPostsQueryResult } from 'config/utils/types';
import type {
	IMdxNode,
	IPost,
	IPostJson,
	IPostBySlugQueryResult,
	IUserQueryResult,
	IUserMetadata
} from '@/types';

export function mdxNodeToPost (mdxNode: IMdxNode): IPost {
	const { frontmatter } = mdxNode;

	if (!frontmatter) {
		throw Error(
			`[mdxNodeToPost] Post missing frontmatter. Post slug: ${
				mdxNode.fields?.slug || 'was not provided'
			}.`
		);
	}

	if (!frontmatter.title) {
		throw Error(
			`[mdxNodeToPost] Post missing title. Post slug: ${
				mdxNode.fields?.slug || 'was not provided'
			}.`
		);
	}

	if (!frontmatter.datePublished) {
		throw Error(
			`[mdxNodeToPost] Post missing publication date. Post slug: ${
				mdxNode.fields?.slug || 'was not provided'
			}.`
		);
	}

	if (!mdxNode.fields) {
		throw Error(
			`[mdxNodeToPost] Post missing fields. Post title: ${frontmatter.title}.`
		);
	}

	if (!mdxNode.fields.slug) {
		throw Error(
			`[mdxNodeToPost] Post missing slug. Post title: ${frontmatter.title}.`
		);
	}

	if (!mdxNode.fields.pathName) {
		throw Error(
			`[mdxNodeToPost] Post missing pathName. Post slug: ${mdxNode.fields.slug}.`
		);
	}

	if (!mdxNode.fields.url) {
		throw Error(
			`[mdxNodeToPost] Post missing url. Post slug: ${mdxNode.fields.slug}.`
		);
	}

	if (!mdxNode.fields.route) {
		throw Error(
			`[mdxNodeToPost] Post missing route. Post slug: ${mdxNode.fields.slug}.`
		);
	}

	if (!mdxNode.timeToRead) {
		throw Error(
			`[mdxNodeToPost] Post missing timeToRead. Post slug: ${mdxNode.fields.slug}.`
		);
	}

	if (!frontmatter.cover) {
		throw Error(
			`[mdxNodeToPost] Post missing cover image. Post slug: ${
				mdxNode.fields?.slug || 'was not provided'
			}.`
		);
	}

	if (!frontmatter.coverAlt) {
		throw Error(
			`[mdxNodeToPost] Post missing cover alt. Post slug: ${
				mdxNode.fields?.slug || 'was not provided'
			}.`
		);
	}

	if (!frontmatter.description) {
		console.warn(
			`[mdxNodeToPost] Post missing description. Post slug: ${
				mdxNode.fields?.slug || 'was not provided'
			}. `
		);
	}

	const tagsFilter = (tag: string | undefined): tag is string =>
		typeof tag !== 'undefined';
	const tagList = frontmatter.tags ? frontmatter.tags.filter(tagsFilter) : [];

	return {
		title: frontmatter.title,

		description: frontmatter.description,
		coverImg: frontmatter.cover.childImageSharp?.gatsbyImageData,
		coverImageUrl: frontmatter.cover.publicURL,
		coverImageAlt: frontmatter.coverAlt,

		datePublished: new Date(frontmatter.datePublished),
		dateModified: new Date(
			frontmatter.dateModified || frontmatter.datePublished
		),

		category: frontmatter.category,
		tags: tagList,

		body: mdxNode.body,
		internalContent: mdxNode.internal?.content,
		excerpt: mdxNode.excerpt,
		timeToRead: mdxNode.timeToRead,

		slug: mdxNode.fields.slug,
		route: mdxNode.fields.route,
		pathName: mdxNode.fields.pathName,
		url: mdxNode.fields.url
	};
}

export function queryToPost (data: IPostBySlugQueryResult): IPost {
	const postData = data.mdx;

	if (!postData) {
		throw Error('[queryToPost]: Query does not contain post data');
	}

	return mdxNodeToPost(postData);
}

export function jsonToPost (meta: IPostJson): IPost {
	const {
		dateModified,
		datePublished,
		slug,
		route,
		pathName,
		url,
		timeToRead,
		title,
		category,
		coverImg,
		coverImageAlt,
		coverImageUrl,
		description,
		excerpt,
		tags,
		relatedPosts
	} = meta;

	return {
		title,

		description,
		coverImg,
		coverImageUrl,
		coverImageAlt,

		datePublished: new Date(datePublished),
		dateModified: new Date(dateModified),

		category,
		tags,

		excerpt,
		timeToRead,

		slug,
		route,
		pathName,
		url,

		relatedPosts: relatedPosts ? relatedPosts.map(jsonToPost) : undefined
	};
}

export function queryToPostsList (res: IGetMdxPostsQueryResult): IPost[] {
	const { edges } = res.allMdx;

	const nodes = edges.map((edge) => edge.node);

	return nodes.map((node) => mdxNodeToPost(node));
}

export function queryToUser (data: IUserQueryResult): IUserMetadata {
	const { user } = data.site.siteMetadata.config;

	if (!user) {
		throw Error('[queryToUser]: Query does not contain user data');
	}

	return user;
}
