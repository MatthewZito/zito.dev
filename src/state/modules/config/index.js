import * as actions from './actions';
import * as getters from './getters';
import * as mutations from './mutations';
import state from './state';

/* Dedicated Vuex Store for App Configurations and Settings */
export default {
  namespaced: true,
  getters,
  mutations,
  actions,
  state
};
