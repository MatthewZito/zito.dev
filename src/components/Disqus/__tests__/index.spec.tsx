import React from 'react';
import { render } from '@testing-library/react';
import cloneDeep from 'clone-deep';
import { mocked } from 'ts-jest/utils';

import { useConfig } from '../../../config';

import { post, config } from '../../../../__tests__/fixtures';

import { DisqusPlugin } from '../index';

jest.mock('../../../config/useConfig', () => ({
	useConfig: jest.fn(() => config)
}));

const mockedUseConfig = mocked(useConfig);

describe('component `DisqusPlugin`', () => {
	it('renders correctly when config.disqusShortname is set', () => {
		const newConfig = cloneDeep(config);
		newConfig.site.disqusShortname = '';

		mockedUseConfig.mockReturnValue(newConfig);

		const { asFragment } = render(<DisqusPlugin post={post} />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('renders correctly when config.disqusShortname is not set', () => {
		const newConfig = cloneDeep(config);
		newConfig.site.disqusShortname = undefined;

		mockedUseConfig.mockReturnValue(newConfig);

		const { asFragment } = render(<DisqusPlugin post={post} />);
		expect(asFragment()).toMatchSnapshot();
	});
});
