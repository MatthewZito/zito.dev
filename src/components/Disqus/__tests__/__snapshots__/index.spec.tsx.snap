// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component \`DisqusPlugin\` renders correctly when config.disqusShortname is not set 1`] = `<DocumentFragment />`;

exports[`component \`DisqusPlugin\` renders correctly when config.disqusShortname is set 1`] = `<DocumentFragment />`;
